api/handlers/videostream.go

```diff
@@ -35,54 +35,340 @@ LICENSE
+		return 0, api.DatastoreReadFailure(ctx)
 	}
-	return ctx.JSON(result)
+
+	// Get a unique ID for the new video stream.
+	key = store.IncompleteKey("VideoStream")
+
+	var err error
+	key, err = store.Put(context.Background(), key, &vs)
+	if err != nil {
+		print(err.Error())
+		return 0, api.DatastoreWriteFailure(ctx)
+	}
+
+	// Return ID of created video stream.
+	return int(key.ID), nil
+
+}
+
+// CreateVideoStream creates a new video stream.
+// BUG: start and end time are required but this is not being enforced.
+// https://github.com/ausocean/openfish/issues/18
+func CreateVideoStream(ctx *fiber.Ctx) error {
+	var body CreateVideoStreamBody
+	err := ctx.BodyParser(&body)
+	if err != nil {
+		return api.InvalidRequestJSON(ctx)
+	}
+
+	// Create video stream entity and add to the datastore.
+	vs := model.VideoStream{
+		StartTime:     body.StartTime,
+		EndTime:       &body.EndTime,
+		StreamUrl:     body.StreamUrl,
+		CaptureSource: body.CaptureSource,
+	}
+	id, err := putVideoStream(ctx, vs)
+	if err != nil {
+		return err
+	}
+
+	// Return ID of created video stream.
+	return ctx.JSON(VideoStreamResult{
+		ID: &id,
+	})
+}
+
+// StartVideoStream creates a new video stream at the current time.
+func StartVideoStream(ctx *fiber.Ctx) error {
+	now := time.Now()
+
+	var body StartVideoStreamBody
+	err := ctx.BodyParser(&body)
+	if err != nil {
+		return api.InvalidRequestJSON(ctx)
+	}
+
+	// Create video stream entity and add to the datastore.
+	vs := model.VideoStream{
+		StartTime:     now,
+		StreamUrl:     body.StreamUrl,
+		CaptureSource: body.CaptureSource,
+	}
+	id, err := putVideoStream(ctx, vs)
+	if err != nil {
+		return err
+	}
+
+	// Return ID of created video stream.
+	return ctx.JSON(VideoStreamResult{
+		ID: &id,
+	})
+}
+
+// EndVideoStream updates the video stream's duration.
+func EndVideoStream(ctx *fiber.Ctx) error {
+	now := time.Now()
+
+	// Parse URL.
+	id, err := strconv.ParseInt(ctx.Params("id"), 10, 64)
+	if err != nil {
+		return api.InvalidRequestURL(ctx)
+	}
+
+	// Update data in the datastore.
+	store := ds_client.Get()
+	key := store.IDKey("VideoStream", id)
+	var videoStream model.VideoStream
+
+	err = store.Update(context.Background(), key, func(e datastore.Entity) {
+		v, ok := e.(*model.VideoStream)
+		if ok {
+			v.EndTime = &now
+		}
+	}, &videoStream)
+	if err != nil {
+		return api.DatastoreReadFailure(ctx)
+	}
+
+	return nil
+}
+
+// UpdateVideoStream updates a video stream.
+func UpdateVideoStream(ctx *fiber.Ctx) error {
+	// Parse URL.
+	id, err := strconv.ParseInt(ctx.Params("id"), 10, 64)
+	if err != nil {
+		return api.InvalidRequestURL(ctx)
+	}
+
+	// Parse body.
+	var body UpdateVideoStreamBody
+	if ctx.BodyParser(&body) != nil {
+		return api.InvalidRequestJSON(ctx)
+	}
+
+	// Update data in the datastore.
+	store := ds_client.Get()
+	key := store.IDKey("VideoStream", id)
+	var videoStream model.VideoStream
+
+	if store.Update(context.Background(), key, func(e datastore.Entity) {
```

Comment:
I still think the call to store.Update should be separated out from the if condition.

Reply:
fixed